vxquery-core/src/main/java/org/apache/vxquery/compiler/rewriter/rules/PushChildIntoDataScanRule.java

```diff
@@ -72,13 +66,17 @@
  * @author prestonc
  */
 public class PushChildIntoDataScanRule extends AbstractUsedVariablesProcessingRule {
-    final StaticContextImpl dCtx = new StaticContextImpl(RootStaticContextImpl.INSTANCE);
+    StaticContext dCtx = null;
     final int ARG_DATA = 0;
     final int ARG_TYPE = 1;
 
     protected boolean processOperator(Mutable<ILogicalOperator> opRef, IOptimizationContext context)
             throws AlgebricksException {
-        AbstractLogicalOperator op1 = (AbstractLogicalOperator) opRef.getValue();;
+        VXQueryOptimizationContext vxqOptCtx = (VXQueryOptimizationContext) context;
+        IMetadataProvider iProvider = (IMetadataProvider) vxqOptCtx.getMetadataProvider();
+        VXQueryMetadataProvider vxqProvider = (VXQueryMetadataProvider) iProvider;
+        dCtx = vxqProvider.getStaticContext();
+        AbstractLogicalOperator op1 = (AbstractLogicalOperator) opRef.getValue();
```

Comment:
I think you could also conditionally set the dCtx when its null. After that it should be static.